libtransact/src/contract/context/key_value.rs

```diff
@@ -0,0 +1,1310 @@
+            .iter()
+            .map(|(key, value)| {
+                StateEntryValueBuilder::new()
+                    .with_key(key.to_string())
+                    .with_value(value.clone())
+                    .build()
+                    .map_err(|err| KeyValueContextError::ProtocolBuildError(Box::new(err)))
+            })
+            .collect::<Result<Vec<StateEntryValue>, KeyValueContextError>>()?;
+        Ok(StateEntryBuilder::new()
+            .with_normalized_key(self.addresser.normalize(key.to_owned()))
+            .with_state_entry_values(state_values)
+            .build()
+            .map_err(|err| KeyValueContextError::ProtocolBuildError(Box::new(err)))?)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use std::sync::{Arc, Mutex};
+
+    use crate::contract::address::{
+        double_key_hash::DoubleKeyHashAddresser, key_hash::KeyHashAddresser,
+        triple_key_hash::TripleKeyHashAddresser,
+    };
+
+    use crate::handler::ContextError;
+
+    /// Create simple state backed by a HashMap.
+    struct TestState {
+        state: HashMap<String, Vec<u8>>,
+    }
+
+    /// Implements the simple state with basic methods to get, set, and delete state values.
+    impl TestState {
+        pub fn new() -> Self {
+            TestState {
+                state: HashMap::new(),
+            }
+        }
+
+        fn get_entries(
+            &self,
+            addresses: &[String],
+        ) -> Result<Vec<(String, Vec<u8>)>, ContextError> {
+            let mut values = Vec::new();
+            addresses.iter().for_each(|key| {
+                if let Some(value) = self.state.get(key) {
+                    values.push((key.to_string(), value.to_vec()))
+                }
+            });
+            Ok(values)
+        }
+
+        fn set_entries(&mut self, entries: Vec<(String, Vec<u8>)>) -> Result<(), ContextError> {
+            entries.iter().for_each(|(key, value)| {
+                match self.state.insert(key.to_string(), value.to_vec()) {
+                    _ => (),
+                }
+            });
+            Ok(())
+        }
+
+        fn delete_entries(&mut self, addresses: &[String]) -> Result<Vec<String>, ContextError> {
+            let mut deleted = Vec::new();
+            addresses.iter().for_each(|key| {
+                if let Some(_) = self.state.remove(key) {
+                    deleted.push(key.to_string());
+                }
+            });
+            Ok(deleted)
+        }
+
+        fn add_receipt_data(&self, _data: Vec<u8>) -> Result<(), ContextError> {
+            Ok(())
+        }
+
+        fn add_event(
+            &self,
+            _event_type: String,
+            _attributes: Vec<(String, String)>,
+            _data: Vec<u8>,
+        ) -> Result<(), ContextError> {
+            Ok(())
+        }
+    }
+
+    /// TestContext using the simple TestState backend to be used as the internal context to
```

Comment:
Should this start with a verb (for consistency)?  See line 30.

Reply:
I reworked these comments to begin with a noun, instead of a verb.